evalml/automl/progress.py

```diff
@@ -0,0 +1,172 @@
+"""Progress abstraction holding stopping criteria and progress information."""
+import logging
+import time
+
+from evalml.utils.logger import get_logger
+
+
+class Progress:
+    """Progress object holding stopping criteria and progress information.
+
+    Args:
+        max_time (int): Maximum time to search for pipelines.
+
+        max_iterations (int): Maximum number of iterations to search.
+
+        max_batches (int): The maximum number of batches of pipelines to search. Parameters max_time, and
+            max_iterations have precedence over stopping the search.
+
+        patience (int): Number of iterations without improvement to stop search early.
+
+        tolerance (float): Minimum percentage difference to qualify as score improvement for early stopping.
+
+        automl_algorithm (str): The automl algorithm to use. Used to calculate iterations if max_batches is selected as stopping criteria.
+
+        objective (str, ObjectiveBase): The objective used in search.
+
+        verbose (boolean): Whether or not to log out stopping information.
+    """
+
+    def __init__(
+        self,
+        max_time=None,
+        max_batches=None,
+        max_iterations=None,
+        patience=None,
+        tolerance=None,
+        automl_algorithm=None,
+        objective=None,
+        verbose=False,
+    ):
+        self.max_time = max_time
+        self.current_time = None
+        self._start_time = None
+        self.max_batches = max_batches
+        self.current_batches = 0
+        self.max_iterations = max_iterations
+        self.current_iterations = 0
+        self.patience = patience
+        self.tolerance = tolerance
+        self.automl_algorithm = automl_algorithm
+        self.objective = objective
+
+        if verbose:
+            self.logger = get_logger(f"{__name__}.verbose")
+        else:
+            self.logger = logging.getLogger(__name__)
+
+    def should_continue(self, results, interrupted=False, mid_batch=False):
```

Comment:
this `mid_batch` flag is actually quite important. If you look at `search()` we call `should_continue` multiple times. In the past, `should_continue` would return false only if the current batch number was greater than `max_batches`. With this flag, we can properly return false when we hit the right batch number.